### DISCUSSION TITLE
ARM Brew or x86 Brew?

### DISCUSSION THREAD

**Discussion by jjvaz25** [Homebrew/discussions#3223] [Getting started] [answered]
I'm on a 2021 Macbook Pro with the M1 Max chip. How do I know if I am using ARM Brew or x86 Brew? Is there anywhere for me to check this?

I installed homebrew via this command:
`/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`

I've been running into all sorts of issues installing packages related to my work and I'm wondering if it is related to the Homebrew that I am using. I was told that Homebrew is distributed in two ways, one that installs arm64 software, one that installs emulated x86 software. Users can run both. How do I know which one I am using?

**Comment by carlocab** [comment, marked as answer]
What is the output of `which -a brew`, `brew config` and `brew doctor`?

**Reply by jjvaz25** [reply]
`which -a brew` output:
`/opt/homebrew/bin/brew`


`brew config` output:
`HOMEBREW_VERSION: 3.4.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: d0a0bbef8db45cca303c607037338d02d8eaa8ca
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 6d59f26a3e8fa552887c3f1c947b2fb0ec55983d
Core tap last commit: 32 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 13.1.6 build 1316
Git: 2.32.0 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.2.1-arm64
CLT: 13.3.1.0.1.1648687083
Xcode: N/A
Rosetta 2: false
`


And lastly `brew doctor` output:
`We found the following "config" scripts:
  /Users/jeff.v/.pyenv/shims/python3.7-config
  /Users/jeff.v/.pyenv/shims/python3.7m-config
  /Users/jeff.v/.pyenv/shims/python-config
  /Users/jeff.v/.pyenv/shims/python3-config
  /Users/jeff.v/.pyenv/shims/pre-commit-validate-config`

**Reply by carlocab** [reply]
You’re using an ARM version of `brew`.

You can tell from any one of the following:
1. Your installation prefix is `/opt/homebrew`
2. The output of `brew config` shows `CPU: 10-core 64-bit arm_firestorm_icestorm`
3. The output of `brew config` shows `Rosetta 2: false`

**Reply by jjvaz25** [reply]
Awesome, thanks for your answer. That solves it for me. Good learning experience as well. Appreciate you taking the time!

**Reply by ghost** [reply]
> You’re using an ARM version of `brew`.
> 
> You can tell from any one of the following:
> 
> 1. Your installation prefix is `/opt/homebrew`
> 2. The output of `brew config` shows `CPU: 10-core 64-bit arm_firestorm_icestorm`
> 3. The output of `brew config` shows `Rosetta 2: false`

Hi @carlocab, I have a similar question, does homebrew automatically check the chip architecture of the machine when installing the package? For example, if the machine is an arm chip, then brew will automatically recognize it, and then only install the arm-adapted software package, not the x86 software package, right?

Does the option `Rosetta 2: false` mean that software that is not compatible with arm will not be installed, and an error will be shown in terminal directly?

**Reply by p-linnane** [reply]
@livrth The supported, default path Homebrew is set up in is determined by the architecture. From the install script:

```
  if [[ "${UNAME_MACHINE}" == "arm64" ]]
  then
    # On ARM macOS, this script installs to /opt/homebrew only
    HOMEBREW_PREFIX="/opt/homebrew"
    HOMEBREW_REPOSITORY="${HOMEBREW_PREFIX}"
  else
    # On Intel macOS, this script installs to /usr/local only
    HOMEBREW_PREFIX="/usr/local"
    HOMEBREW_REPOSITORY="${HOMEBREW_PREFIX}/Homebrew"
  fi
```

`Rosetta 2: false` just indicates that `brew config` was run with using Rosetta. If you were to run `arch -x86_64 brew config` it would return `Rosetta 2: true`

**Reply by carlocab** [reply]
> does homebrew automatically check the chip architecture of the machine when installing the package?

Yes.

> For example, if the machine is an arm chip, then brew will automatically recognize it, and then only install the arm-adapted software package, not the x86 software package, right?

Not exactly. On an M1 machine, `brew install --formula` will install M1-native software as long as your prefix is not `/usr/local`. If your prefix is `/usr/local`, `brew install --formula` will install software built for an Intel machine. Casks can install anything, so there are no guarantees about casks.

> Does the option `Rosetta 2: false` mean that software that is not compatible with arm will not be installed, and an error will be shown in terminal directly?

`Rosetta 2: false` means that `brew` is not running under the Rosetta translation layer at the moment you called it.

**Reply by LronDC** [reply]
@carlocab Hi, thanks for your explanation. I have a further question. How to change an x86 brew to an arm brew easily? Do I need to uninstall x86 brew and install a new one?

**Reply by gromgit** [reply]
> @carlocab Hi, thanks for your explanation. I have a further question. How to change an x86 brew to an arm brew easily? Do I need to uninstall x86 brew and install a new one?

No, you simply run the Homebrew installer in native mode, i.e. not under Rosetta. That puts a brand-new ARM-based Homebrew hierarchy under `/opt/homebrew`, completely separate from the x86 one under `/usr/local`.

Then it's just a matter of managing your `PATH` properly:
* `eval "$(/opt/homebrew/bin/brew shellenv)"` makes ARM Homebrew the default
* `eval "$(/usr/local/bin/brew shellenv)"` makes x86 Homebrew the default

And regardless of what your current default is, you can always manually run `/usr/local/bin/brew <subcommand>...` or `/opt/homebrew/bin/brew <subcommand>...` to directly operate on x86 and ARM Homebrew respectively.

**Comment by KunalMaverick** [comment]
@carlocab 
I have a very peculiar problem XD


bash-3.2$ brew config
HOMEBREW_VERSION: 4.1.13
ORIGIN: https://github.com/Homebrew/brew
HEAD: a8519f78fb63f2f2266950bdd8141037da69f8bd
Last commit: 2 days ago
Core tap JSON: 27 Sep 17:31 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_REPOSITORY: /opt/homebrew
HOMEBREW_CELLAR: /opt/homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit westmere
Clang: 14.0.0 build 1400
Git: 2.39.2 => /opt/homebrew/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.6.2-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: true


looking at the path, rosetta should have been false

what exactly is happening??

**Reply by SMillerDev** [reply]
Your Homebrew version is very old and you're on an unsupported version of macOS. But the reason you're running in Rosetta is probably because your terminal emulator is not a native ARM build.
